Step1 수정요청사항 반영해서 올립니다. 리뷰부탁드립니다.

jinyounglee opened the pull request.

jinyounglee committed:
feat: step1 구현

Commit:
refactor: 메서드 분리, 테스트 코드 작성

Commit:
refactor: 리뷰 요청사항 반영
- 사용하지 않는 validate 클래스 제거
- Question 삭제시 사용자 비교 메서드 추출 및 equals 재정의
- validateAnswer 를 delete 메서드 내로 이동
- 불필요한 @Transactional 어노테이션 제거
- Question 삭제시 History 처리 부분 domain 영역으로 이동

Commit:
refactor: 리뷰 요청사항 반영
- 사용하지 않는 validate 클래스 제거
- Question 삭제시 사용자 비교 메서드 추출 및 equals 재정의
- validateAnswer 를 delete 메서드 내로 이동
- 불필요한 @Transactional 어노테이션 제거
- Question 삭제시 History 처리 부분 domain 영역으로 이동

nooose commented on src/main/java/nextstep/qna/service/QnAService.java:
```diff
@@ -26,24 +23,18 @@ public class QnAService {
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
-        deleteHistoryService.saveAll(deleteHistories);
+        DeleteHistories deleteHistories = question.delete(loginUser);
+
+        Answers answers = question.getAnswers();
+        answers.deleteAnswers(loginUser);
+
+        deleteHistories.addAllDeleteHistories(answers.deleteAllHistories());
+        addDeleteHistories(deleteHistories);
```
Question 멤버 변수를 봤을 때 answers를 가지고 있습니다.
즉, Question 내부에서 answers를 호출할 수 있다는 의미인데요

`question.delete` 메서드 내부적으로 answers에게 `delete`를 요청해 볼 수 있을 것 같아요

또한 `DeleteHistories deleteHistories = question.delete(loginUser);` 와 같이
`answers.delete`의 반환타입을 DeleteHistories로 정의할 수 있을 것 같습니다.

question 도메인에게 직접 메시지를 전달하면 question의 삭제 히스토리와 answers의 삭제 히스토리를 한 번에 반환받을 수 있도록 
리팩토링 해보시면 좋겠습니다.

유명한 예시로 설명을 해볼게요
진영님이 카페를 가서 커피를 주고 샀을 때 점원에게
**_돈을 받고 물을 끓인 후 커피를 내려주고 나한테 다시 줘_** 라는 행위를 시키지는 않잖아요?
진영님은 단순히 점원에게 커피를 달라는 메시지만 전달할 텐데요

question의 delete도 마찬가지로 삭제 메시지를 보내면
내부적으로 삭제 처리를 하고 삭제 히스토리를 반환받을 수 있도록 설계해 볼 수 있을 것 같습니다.

내용이 어렵다면 [이전 PR](#270) 에 남긴 코멘트도 다시 한번 읽어보시면 도움이 될 것 같아요

nooose replied:
서비스에서 question으로 부터 값을 꺼내와서 하나하나 로직을 풀지않고
```java
Question question = questionRepository.findById(questionId).orElseThrow(NotFoundException::new);
DeleteHistories deleteHistories = question.delete(loginUser);
deleteHistoryService.saveAll(deleteHistories.getDeleteHistories());
```
이런 흐름이 나오도록 변경해 보시면 좋겠습니다.

jinyounglee replied:
answer와 question 삭제를 따로 해주는데 history는 한번에 받아와서 처리해주는게 맞을까요????

nooose commented on src/main/java/nextstep/qna/domain/DeleteHistories.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class DeleteHistories {
```
요구사항에 따라 일급컬렉션으로 구현하신 것 같습니다.
DeleteHistories를 봤을 때 단순 `List<DeleteHistory>`를 사용하는 것과 차이가 없을 것 같아요
Answers 같이 로직을 갖는 상황이라면 장점이 될 수 있지만,
DeleteHistories에서는 오버헤드만 늘어나게 되어 불필요한 것 같습니다.

항상 트레이드오프 관계가 존재하니 참고만 해주세요

nooose reviewed (changes requested):
안녕하세요 진영님
리뷰가 늦어서 정말 죄송합니다.

코멘트 남겼는데요 읽어보시고 재요청 부탁드립니다.
질문 있으시면 언제든지 DM 남겨주세요~

jinyounglee committed:
fix: Question, Answer 삭제 History를 한번에 받아와 처리하도록 수정

nooose commented on src/main/java/nextstep/qna/service/QnAService.java:
```diff
@@ -26,24 +23,13 @@ public class QnAService {
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
-        deleteHistoryService.saveAll(deleteHistories);
+        DeleteHistories deleteHistories = question.delete(loginUser);
+
+        Answers answers = question.getAnswers();
+        answers.deleteAnswers(loginUser);
```
```suggestion
        DeleteHistories deleteHistories = question.delete(loginUser);
```

```java

        Answers answers = question.getAnswers();
        answers.deleteAnswers(loginUser);
```
이 부분도 `question.delete` 안에서 처리할 수 있지 않을까요?

nooose replied:
```java
DeleteHistories deleteHistories = question.delete(loginUser);
        Answers answers = question.getAnswers();
        answers.deleteAnswers(loginUser);
```

삭제 히스토리 처리와 삭제 처리를 각각 별도 호출을 통해 처리하고 있습니다.
삭제가 되지 않았는데 삭제 히스토리를 만들 수 있다는 것은 어쩌면 이상하지 않을까요?

answers도 삭제와 동시에 삭제 히스토리를 반환하도록 만들었다면 
서비스에서 deleteAnswers를 호출할 필요가 사라질 것 같고 버그 발생도 줄일 수 있을 것 같아요

nooose commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -81,12 +88,41 @@ public boolean isDeleted() {
+    }
+
+    private void validateDelete(NsUser loginUser) throws CannotDeleteException {
+        if (!Objects.equals(loginUser, writer)) {
```
```suggestion
        if (!isOwner(loginUser)) {
```
이렇게 if 안 조건을 메서드를 분리하여 가독성있게 변경할 수 있을 것 같습니다.

nooose commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -81,12 +88,41 @@ public boolean isDeleted() {
     }
+
+    public DeleteHistories delete(NsUser loginUser) throws CannotDeleteException {
+        validateDelete(loginUser);
+        this.deleted = true;
+        deleteAnswers(loginUser);
+        return makeDeletehistories();
```
delete 메소드 내 private 메소드 호출 순서가 다음과 같네요 
1. validateDelete
2. deleteAnswers
3. makeDeletehistories

메소드 배치도 위 순서에 맞게 재배치하면 좋을 것 같아요
이러면 코드를 위에서 아래로 읽어나갈 수 있게되어 코드 가독성이 좋아집니다.

```java

    private void validateDelete(NsUser loginUser) throws CannotDeleteException {
        ...
    }
    
    private void deleteAnswers(NsUser loginUser) throws CannotDeleteException {
        ...
    }   

    private DeleteHistories makeDeletehistories() {
        ...
    }

```

nooose replied:
아래와 같이 작성한다면 서비스에서 delete 메서드 하나만 호출해도 되겠네요
```suggestion
        validateDelete(loginUser);
        this.deleted = true;
        답변_삭제_히스토리 = answers.delete(loginUser);
        return makeDeletehistories(답변_삭제_히스토리);
```

nooose reviewed (changes requested):
안녕하세요 진영님!
좀 더 개선할 수 있을 것 같아요
코멘트 읽어보시고 재요청 부탁드립니다.

jinyounglee committed:
refactor:
- 삭제 처리 기능 Question delete 메서드 안으로 이동
- private method 호출 순서에 맞게 위치 변경
- Deletehistories 클래스 제거

nooose commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -97,32 +94,34 @@ public String toString() {
+    private List<DeleteHistory> deleteAnswers(NsUser loginUser) throws CannotDeleteException {
         if (Objects.nonNull(this.answers)) {
             this.answers.deleteAnswers(loginUser);
+            return this.answers.deleteAllHistories();
```
메소드 이름에 answers라는 이름은 중복적인 의미일 것 같아요
Question의 delete처럼 answers에게 delete 메시지만 전달해도 될 것 같네요

nooose replied:
```java
return this.answers.delete(); 
```

Question의 delete 처럼 answers의 delete도 DeleteHistory를 반환하도록 변경할 수 있지 않을까요?

nooose commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -81,12 +85,43 @@ public boolean isDeleted() {
+    }
+
+    private List<DeleteHistory> deleteAnswers(NsUser loginUser) throws CannotDeleteException {
+        if (Objects.nonNull(this.answers)) {
```
```suggestion
        if (answers != null) {
```
null 체크를 동등연산자를 사용해서 체크할 수 있었을텐데
Objects의 `nonNull` 메소드를 사용하셨네요

<img width="636" alt="image" src="https://github.com/next-step/java-lms/assets/36617783/40e4e74c-d291-48f0-8a82-7f3f1265379a">

문서를 보시면 `This method exists to be used as a java.util.function.Predicate, filter(Objects::nonNull)`
람다 표현식을 간단하게 표시할 때 사용하라고 만든 메서드입니다.
null을 직접 체크하는 방법도 의미전달이 충분히 되기 때문에 굳이 메소드 콜을 한번더 할 필요는 없을 것 같다는 생각이 듭니다.

nooose reviewed (approved):
안녕하세요 진영님!
코멘트 남겼는데 읽어보시고 다음 단계 진행해 주세요
피드백 반영하시느라 정말 수고 많으셨습니다.

nooose commented:
메서드 배치 순서에 관련된 피드백은 지켜지지 않은 것 같네요 😢
